### PR TITLE
ua: ua_event for DnD rejected call

### PR DESCRIPTION
**Pull request by cspiel1**
Adds a UA event for rejected incoming calls due to activated DnD.

The UA event contains
- Contact header: Can be used by the application in order to call back.
- Display name of the From header: Can be used by the application to display the peer of the rejected call.
- From Uri: May also be used to display the peer.

### CODE REVIEW

**Comment by cspiel1** [comment]
With module ctrl_dbus the tool dbus-monitor shows this DBUS signal:
```
signal time=1712060692.114381 sender=:1.7928 -> destination=(null destination) serial=7 path=/baresip; interface=com.github.Baresip; member=event
   string "call"
   string "CALL_REJECTED"
   string "{"type":"CALL_REJECTED","class":"call","accountaor":"sip:A@localhost","param":"{\"reason\":\"DnD\",\"contact\":\"<sip:od5-0xae0988@192.168.110.182:5060>\",\"display\":\"OD5\",\"from\":\"sip:od5@192.168.110.182\"}"}"
```

**Comment by cspiel1** [comment]
The current solution introduces a new event type `UA_EVENT_CALL_REJECTED` which has the event class "call", but we don't want to allocate a call object only for encoding the UA event. So we pass `NULL` as call pointer and encode the wanted data into the `prm` parameter of the UA event handlers.

If someone has a better idea, please let me know! We think that the Contact header + From header + Display name is useful for rejected calls. At least the Contact header currently is not part of real call UA events.

TODOs:
- Decode+encode the Contact header for security reason.
- I'll try to use the JSON API of re for encoding. Maybe it looks better.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> The current solution introduces a new event type
> `UA_EVENT_CALL_REJECTED` which has the event class "call", but we
> don't want to allocate a call object only for encoding the UA
> event. So we pass `NULL` as call pointer and encode the wanted data
> into the `prm` parameter of the UA event handlers. 
> 
> If someone has a better idea, please let me know! We think that the
> Contact header + From header + Display name is useful for rejected
> calls. At least the Contact header currently is not part of real call
> UA events.

Why don't you implement DnD in the app (menu module)?

**Comment by cspiel1** [comment]
The DnD feature is already in the baresip core. It is checked before a call is allocated and `call_accept()` is called. The 480 "Temporarily Unavailable" is sent before call allocation, like
- 404 "Not Found"
- 486 "Max Calls"
- 420 "Bad Extension"
- 421 "Extension required"

I think we should not change this behavior.

**Comment by cspiel1** [comment]
I tried `odict_alloc`, and `odict_entry_add` and I think that it does make it better. An `odict_entry_add` with type `struct pl` is missing.

Anyway, I add a commit that corrects the escaping of the inner JSON object.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> The DnD feature is already in the baresip core. It is checked before a call is allocated and `call_accept()` is called. The 480 "Temporarily Unavailable" is sent before call allocation, like
> - 404 "Not Found"
> - 486 "Max Calls"
> - 420 "Bad Extension"
> - 421 "Extension required"
> 
> I think we should not change this behavior.

Keep the behavior, but don't use it.  Instead, as I mentioned, implement
it in the app per account. which is a much better choice.

**Comment by cspiel1** [comment]
Maybe this would be possible. I have to think about it. For now I can say, with the solution you propose we get these unwanted behavior if the device is in DnD mode:
- the call is allocated,
- `call_accept()` is called,
- 180 Ringing is sent back,
- the UA_EVENT_CALL_INCOMING event is sent out and these will pass on our devices through
  - intercom.so
  - multicast.so
  - menu.so

The UA_EVENT_CALL_INCOMING will result in unwanted behavior for the user. And on the peer there will be also unwanted behavior.

And if we add a new application module with another UA event handler that is processed first, we still have the problem that 180 Ringing is sent to the peer.

**Comment by cspiel1** [comment]
When visualizing the new UA event with autotest, the JSON looks correct. Thus I guess that unwanted backslashes are printed by the dbus-monitor tool.

```
autotest: [ ua=sip:A@localhost call= ] event: CALL_REJECTED ({"reason":"DnD","contact":"sip:od5-0xae0988@192.168.110.182:5060","display":"OD5","from":"sip:od5@192.168.110.182"})
```

I set this to "Ready for review".

**Comment by juha-h** [comment]
Christian Spielberger writes:

> Maybe this would be possible. I have to think about it. For now I can
> say, with the solution you propose we get these unwanted behavior if
> the device is in DnD mode: 
> - the call is allocated,
> - `call_accept()` is called,
> - 180 Ringing is sent back,
> - the UA_EVENT_CALL_INCOMING event is sent out and these will pass on our devices through

When call comes in, why is 180 Ringing sent back before the app has had
a chance to decide what to do with the call?  In my opinion, the app
should get CALL_INCOMING event and then be able to response with 180
Ringing, 486 Busy Here, ..., or 200 Answering.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> And if we add a new application module with another UA event handler
> that is processed first, we still have the problem that 180 Ringing is
> sent to the peer.

That exactly is the problem.  There should be no automatic 180 Ringing
response.  For example, my app accepts only one simultaneous call.  When
the second call comes in, my app responds with 486 Busy Here.  Now that
I checked with SIP trace, there was unwanted 180 Ringing before the 486:

...
04-02 17:44:41.946  6471  6549 D Baresip Lib: call: alloc with params laddr=10.0.0.2, af=AF_INET, use_rtp=1
04-02 17:44:41.946  6471  6549 D Baresip Lib: call: use_video=0
04-02 17:44:41.953  6471  6549 D Baresip Lib: 17:44:41.953#
04-02 17:44:41.954  6471  6549 D Baresip Lib: TLS 10.0.0.2:60304 -> xxx.xxx.xxx.xxx:yyyy
04-02 17:44:41.954  6471  6549 D Baresip Lib: SIP/2.0 180 Ringing
...
04-02 17:44:41.957  6471  6548 D Baresip Lib: ua event (CALL_INCOMING)
***@***.***
...
04-02 17:44:41.977  6471  6549 D Baresip Lib: call: rejecting incoming call from ***@***.*** (486 Busy Here)
...

There must be some way to prevent the automatic 180 Ringing response.

**Comment by juha-h** [comment]
Other examples include: user does not want to accept calls if caller is not in contacts or if caller is in spammers.  In all these cases, there should not be a 180 Ringing response.

**Comment by cspiel1** [comment]
That makes sense for me. So I put this PR back to draft and I try to make a new one for further discussion.

What about the 183 Session Progress? Baresip sends always
1. 180 Ringing
2. 183 Session Progress

so both in this order.

File: src/ua.c [hunk 1]
```diff
@@ -704,6 +704,8 @@ void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
 	if (uag_dnd()) {
 		(void)sip_treply(NULL, uag_sip(), msg,
 			480,"Temporarily Unavailable");
+
+		ua_event_rejected(ua, "DnD", msg);
```
Comment (alfredh) [inline src/ua.c:708]: Can we move this to menu module ?

The less logic in the core, the better ...

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> Can we move this to menu module ?

That was my suggestion too.

> The less logic in the core, the better ...

Yes.

**Comment by juha-h** [comment]
When call comes in and user does not want/is not able to receive it, use might want to response with `302 Moved Temporarily` to somewhere, e.g., to voicemail, but `ua_hangup` does not have contact URI argument.

**Comment by cspiel1** [comment]
Currently the state `CALL_STATE_INCOMING` and the event `CALL_EVENT_INCOMING` mean that the call was already answered with 180 "Ringing". There are already several modules (maybe external apps) that depend on this.
- module `menu`
- module `gtk`
- modules in baresip apps

Should we add a state/event `PRE_INCOMING`?

Better wording would be `INCOMING` and `ACCEPTED`, but that would probably break applications.

**Comment by juha-h** [comment]
How about CALL_[STATE|EVENT]_ARRIVED?

**Comment by cspiel1** [comment]
Will be replaced. DnD moved to module menu.

**Pull request closed without merging**